src/services/organizeImports.ts

```diff
@@ -0,0 +1,291 @@
+/* @internal */
+namespace ts.OrganizeImports {
+    export function organizeImports(
+        sourceFile: SourceFile,
+        formatContext: formatting.FormatContext,
+        host: LanguageServiceHost,
+        cancellationToken: CancellationToken) {
+
+        // All of the (old) ImportDeclarations in the file, in syntactic order.
+        const oldImportDecls: ImportDeclaration[] = [];
+
+        forEachChild(sourceFile, node => {
+            cancellationToken.throwIfCancellationRequested();
+            if (isImportDeclaration(node)) {
+                oldImportDecls.push(node);
+            }
+            // TODO (https://github.com/Microsoft/TypeScript/issues/10020): sort *within* ambient modules (find using isAmbientModule)
+        });
+
+        if (oldImportDecls.length === 0) {
+            return [];
+        }
+
+        const usedImportDecls = removeUnusedImports(oldImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+        const sortedImportDecls = sortImports(usedImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+        const coalescedImportDecls = coalesceImports(sortedImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+
+        // All of the (new) ImportDeclarations in the file, in sorted order.
+        const newImportDecls = coalescedImportDecls;

```

Comment:
Are locals expensive in TS?  I like naming things - it makes code easier to read.

Reply:
I'm sure modern JS engines know that these are the same thing. I normally agree that giving a descriptive name for a variable is useful, but `const coalescedImportDecls = coalesceImports(sortedImportDecls);` just repeats the same information in two different names, since the function name already tells you what it's doing -- same for the above lines.

Reply:
In the debugger, it's easy to inspect each stage individually.  If you'd prefer not to have any redundant text, arguably they should be combined into a single expression with only final value stored in a named variable.

Reply:
For pure functions, you can enter e.g. `coalesceImports(sortedImportDecls)` into the watch list for inspection without needing a named local variable.